# Coding Exercises (Part 8)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Advanced Topics

### Exercise 8: Creating and Removing Rows and Columns

In [1]:
#run the cell!
import pandas as pd
pd.options.display.float_format = '{:.2f}'.format

In [2]:
#run the cell!
listings = pd.read_csv("listings_clean.csv")

In [3]:
#run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0      1999     Capital Goods   
1      2016  Basic Industries   
2      1969        Technology   
3      2014       Health Care   
4      1969    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [4]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
Symbol        6852 non-null object
Exchange      6852 non-null object
Name          6852 non-null object
Last_Price    6852 non-null float64
Market_Cap    5954 non-null float64
IPO_Year      6852 non-null int64
Sector        5309 non-null object
Industry      5309 non-null object
dtypes: float64(2), int64(1), object(5)
memory usage: 428.3+ KB


76. __Drop__ the columns __"Symbol"__ and __"Sector"__ from the DataFrame listings! __Save__ the change! __Fill in the gaps!__

In [5]:
listings.drop(columns = ["Symbol", "Sector"], inplace = True)

In [6]:
# run the cell!
listings.head()

Exchange                           Name  Last_Price     Market_Cap  \
0     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year                                          Industry  
0      1999  Biotechnology: Laboratory Analytical Instruments  
1      2016                                          Aluminum  
2      1969                                      EDP Services  
3      2014                              Medical Specialities  
4      1969                     Air Freight/Delivery Services

77. __Drop__ all rows/listings on __"AMEX"__ from the DataFrame listings! Use the __loc__ operator and __overwrite listings__!<br> __Fill in the gaps!__

In [7]:
listings = listings.loc[listings.Exchange != "AMEX"]

In [8]:
# run the cell!
listings.head()

Exchange                           Name  Last_Price     Market_Cap  \
0     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year                                          Industry  
0      1999  Biotechnology: Laboratory Analytical Instruments  
1      2016                                          Aluminum  
2      1969                                      EDP Services  
3      2014                              Medical Specialities  
4      1969                     Air Freight/Delivery Services

In [9]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6541 entries, 0 to 6851
Data columns (total 6 columns):
Exchange      6541 non-null object
Name          6541 non-null object
Last_Price    6541 non-null float64
Market_Cap    5677 non-null float64
IPO_Year      6541 non-null int64
Industry      5085 non-null object
dtypes: float64(2), int64(1), object(3)
memory usage: 357.7+ KB


78. __Add__ the new column __MCAP_MUSD__ (Market Cap in Million USD) to the DataFrame listings by dividing Market_Cap by 1 million. __Fill in the gaps__!

In [10]:
listings["MCAP_MUSD"] = listings.Market_Cap/1000000

__Inspect__! What is the __MCAP_MUSD__ for AAC Holdings?

In [11]:
# run the cell!
listings.head()

Exchange                           Name  Last_Price     Market_Cap  \
0     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year                                          Industry  MCAP_MUSD  
0      1999  Biotechnology: Laboratory Analytical Instruments   25934.70  
1      2016                                          Aluminum    5407.81  
2      1969                                      EDP Services   42781.13  
3      2014                              Medical Specialities      53.14  
4      1969                     Air Freight/Delivery Services   15276.87

AAC Holdings has a Market Cap of ... 53.14 MUSD!

79. __Add__ the new column __Shares__ with the total amount of __shares outstanding__ for each listing. __Divide__ __Market_Cap__ by the __Last_Price__! <br>
Instead of using the arithmetic sign "/", use an __appropriate method for division__! 

In [12]:
listings["Shares"] = listings.Market_Cap.div(listings.Last_Price)

In [13]:
# run the cell!
listings.head()

Exchange                           Name  Last_Price     Market_Cap  \
0     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year                                          Industry  MCAP_MUSD  \
0      1999  Biotechnology: Laboratory Analytical Instruments   25934.70   
1      2016                                          Aluminum    5407.81   
2      1969                                      EDP Services   42781.13   
3      2014                              Medical Specialities      53.14   
4      1969                     Air Freight/Delivery Services   15276.87   

        Shares  
0 317515869.00  
1 185516624.00  
2 567464270.00  
3  24602355.00  
4 449055548.00

In [14]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6541 entries, 0 to 6851
Data columns (total 8 columns):
Exchange      6541 non-null object
Name          6541 non-null object
Last_Price    6541 non-null float64
Market_Cap    5677 non-null float64
IPO_Year      6541 non-null int64
Industry      5085 non-null object
MCAP_MUSD     5677 non-null float64
Shares        5583 non-null float64
dtypes: float64(4), int64(1), object(3)
memory usage: 459.9+ KB


80. __Sort__ listings by the __Shares__ column from __high to low__. Which listing has the __largest number__ of shares outstanding

In [15]:
listings.sort_values(by = "Shares", ascending= False)

Exchange                                             Name  Last_Price  \
3812     NYSE                         Lloyds Banking Group Plc        3.28   
5422     NYSE                            Banco Santander, S.A.        4.91   
25       NYSE                                       Ambev S.A.        4.54   
4169     NYSE               Mitsubishi UFJ Financial Group Inc        5.04   
3929     NYSE                     Mizuho Financial Group, Inc.        3.12   
3349     NYSE                   Itau Unibanco Banco Holding SA        8.90   
656      NYSE                                Banco Bradesco Sa        9.34   
625      NYSE                      Bank of America Corporation       29.07   
2610     NYSE                         General Electric Company        9.16   
4109   NASDAQ                            Microsoft Corporation      120.19   
938      NYSE                        Banco Santander Brasil SA       11.76   
5916     NYSE                                        AT&T Inc.       31.88   
5645     NYSE              Sumitomo Mitsui Financial Group Inc        7.14   
658      NYSE                                Banco Bradesco Sa        8.18   
674      NYSE             Banco Bilbao Viscaya Argentaria S.A.        6.01   
4746     NYSE              Petroleo Brasileiro S.A.- Petrobras       16.99   
5182     NYSE                    Royal Bank Scotland plc (The)        6.76   
6661     NYSE                                    Wipro Limited        4.30   
3667     NYSE             China Life Insurance Company Limited       13.73   
4404     NYSE                                Nokia Corporation        5.85   
4827     NYSE                                     Pfizer, Inc.       42.73   
6004     NYSE                                    Telefonica SA        8.35   
6207     NYSE  Taiwan Semiconductor Manufacturing Company Ltd.       42.37   
6414     NYSE                                        VALE S.A.       13.56   
5593   NASDAQ                          Sirius XM Holdings Inc.        5.99   
11     NASDAQ                                       Apple Inc.      200.62   
6626     NYSE                            Wells Fargo & Company       47.79   
1354   NASDAQ                              Comcast Corporation       41.08   
3265   NASDAQ                                Intel Corporation       55.75   
1537   NASDAQ                              Cisco Systems, Inc.       55.82   
...       ...                                              ...         ...   
6630     NYSE                            Wells Fargo & Company       24.55   
6631     NYSE                            Wells Fargo & Company       26.00   
6632     NYSE                            Wells Fargo & Company       27.81   
6633     NYSE                            Wells Fargo & Company       25.59   
6634     NYSE                            Wells Fargo & Company       26.10   
6635     NYSE                            Wells Fargo & Company       25.45   
6636     NYSE                            Wells Fargo & Company       25.27   
6637     NYSE                            Wells Fargo & Company       25.49   
6638     NYSE                            Wells Fargo & Company       25.67   
6644   NASDAQ                         WhiteHorse Finance, Inc.       25.84   
6650   NASDAQ       Wheeler Real Estate Investment Trust, Inc.        0.01   
6656   NASDAQ          Western Asset Short Duration Income ETF       25.45   
6690     NYSE                      Washington Prime Group Inc.       21.54   
6691     NYSE                      Washington Prime Group Inc.       19.82   
6697     NYSE                         W.R. Berkley Corporation       25.25   
6698     NYSE                         W.R. Berkley Corporation       25.74   
6699     NYSE                         W.R. Berkley Corporation       25.17   
6700     NYSE                         W.R. Berkley Corporation       25.27   
6706   NASDAQ                        Pensare Acquisition Corp.        0.28   
6707   NASDAQ              

Lloyds Banking Group has over 17 billion shares outstanding!

81. Let´s assume we have the listings for NYSE, NASDAQ and AMEX in __three separate datasets__. __Import__ the three datasets into Pandas and __inspect__!

In [16]:
# run the cell!
nyse = pd.read_csv("NYSE.csv")
nasdaq = pd.read_csv("NASDAQ.csv") 
amex = pd.read_csv("AMEX.csv") 

In [17]:
# run the cell!
nyse.head()

Symbol                    Name  LastSale       MarketCap  ADR TSO  IPOyear  \
0    DDD  3D Systems Corporation     10.91   1272949528.47      nan      nan   
1    MMM              3M Company    212.11 122153435461.96      nan      nan   
2   WBAI         500.com Limited     13.85    588423939.55      nan  2013.00   
3   WUBA             58.com Inc.     70.11  10391864681.79      nan  2013.00   
4   EGHT                 8x8 Inc     21.09   2017357580.82      nan      nan   

              Sector                                         Industry  
0         Technology          Computer Software: Prepackaged Software  
1        Health Care                       Medical/Dental Instruments  
2  Consumer Services            Services-Misc. Amusement & Recreation  
3         Technology  Computer Software: Programming, Data Processing  
4         Technology                                     EDP Services

In [18]:
# run the cell!
nyse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3106 entries, 0 to 3105
Data columns (total 8 columns):
Symbol       3106 non-null object
Name         3106 non-null object
LastSale     3015 non-null float64
MarketCap    3106 non-null float64
ADR TSO      0 non-null float64
IPOyear      1416 non-null float64
Sector       2121 non-null object
Industry     2121 non-null object
dtypes: float64(4), object(4)
memory usage: 194.2+ KB


In [19]:
# run the cell!
nasdaq.head()

Symbol                                    Name  LastSale     MarketCap  \
0     YI                               111, Inc.      7.10   56690340.50   
1    PIH  1347 Property Insurance Holdings, Inc.      5.20   31266372.80   
2  PIHPP  1347 Property Insurance Holdings, Inc.     24.90   17430000.00   
3   TURN                180 Degree Capital Corp.      1.91   59286575.61   
4   FLWS                 1-800 FLOWERS.COM, Inc.     18.81 1206817397.28   

     ADR TSO  IPOyear             Sector                    Industry  
0 7984555.00  2018.00        Health Care    Medical/Nursing Services  
1        nan  2014.00            Finance  Property-Casualty Insurers  
2        nan      nan            Finance  Property-Casualty Insurers  
3        nan      nan            Finance  Finance/Investors Services  
4        nan  1999.00  Consumer Services      Other Specialty Stores

In [20]:
# run the cell!
nasdaq.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3435 entries, 0 to 3434
Data columns (total 8 columns):
Symbol       3435 non-null object
Name         3435 non-null object
LastSale     3432 non-null float64
MarketCap    3435 non-null float64
ADR TSO      140 non-null float64
IPOyear      1599 non-null float64
Sector       2964 non-null object
Industry     2964 non-null object
dtypes: float64(4), object(4)
memory usage: 214.8+ KB


In [21]:
# run the cell!
amex.head()

Symbol                                               Name  LastSale  \
0   XXII                            22nd Century Group, Inc      1.89   
1    FAX              Aberdeen Asia-Pacific Income Fund Inc      4.14   
2    IAF                 Aberdeen Australia Equity Fund Inc      5.45   
3    AEF  Aberdeen Emerging Markets Equity Income Fund, ...      7.58   
4    FCO                  Aberdeen Global Income Fund, Inc.      8.16   

      MarketCap  ADR TSO  IPOyear                 Sector  \
0  235574500.77      nan      nan  Consumer Non-Durables   
1 1042993611.36      nan  1986.00                    NaN   
2  123945676.70      nan      nan                    NaN   
3  453154667.63      nan      nan                    NaN   
4   71194278.24      nan  1992.00                    NaN   

                Industry  
0  Farming/Seeds/Milling  
1                    NaN  
2                    NaN  
3                    NaN  
4                    NaN

In [22]:
# run the cell!
amex.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311 entries, 0 to 310
Data columns (total 8 columns):
Symbol       311 non-null object
Name         311 non-null object
LastSale     298 non-null float64
MarketCap    311 non-null float64
ADR TSO      0 non-null float64
IPOyear      90 non-null float64
Sector       224 non-null object
Industry     224 non-null object
dtypes: float64(4), object(4)
memory usage: 19.5+ KB


82. __Concatenate__ the three DataFrames into one DataFrame and save the new DataFrame in the variable __all_listings__! <br>
Each row/listing should contain information about the respective __Exchange__ (__NYSE__, __NASDAQ__ or __AMEX__)! __Fill in the gaps__!

In [23]:
all_listings = pd.concat([nyse, nasdaq, amex], keys = ["NYSE", "NASDAQ", "AMEX"], names = ["Exchange"])

In [24]:
# run the cell!
all_listings.head()

Symbol                    Name  LastSale       MarketCap  ADR TSO  \
Exchange                                                                       
NYSE     0    DDD  3D Systems Corporation     10.91   1272949528.47      nan   
         1    MMM              3M Company    212.11 122153435461.96      nan   
         2   WBAI         500.com Limited     13.85    588423939.55      nan   
         3   WUBA             58.com Inc.     70.11  10391864681.79      nan   
         4   EGHT                 8x8 Inc     21.09   2017357580.82      nan   

            IPOyear             Sector  \
Exchange                                 
NYSE     0      nan         Technology   
         1      nan        Health Care   
         2  2013.00  Consumer Services   
         3  2013.00         Technology   
         4      nan         Technology   

                                                   Industry  
Exchange                                                     
NYSE     0          Computer Software: Prepackaged Software  
         1                       Medical/Dental Instruments  
         2            Services-Misc. Amusement & Recreation  
         3  Computer Software: Programming, Data Processing  
         4                                     EDP Services

In [25]:
# run the cell to tidy up!
all_listings = all_listings.reset_index().drop(columns = "level_1")

How many listings does __listings_all__ contain in total?

In [26]:
# run and inspect!
all_listings.head()

Exchange Symbol                    Name  LastSale       MarketCap  ADR TSO  \
0     NYSE    DDD  3D Systems Corporation     10.91   1272949528.47      nan   
1     NYSE    MMM              3M Company    212.11 122153435461.96      nan   
2     NYSE   WBAI         500.com Limited     13.85    588423939.55      nan   
3     NYSE   WUBA             58.com Inc.     70.11  10391864681.79      nan   
4     NYSE   EGHT                 8x8 Inc     21.09   2017357580.82      nan   

   IPOyear             Sector                                         Industry  
0      nan         Technology          Computer Software: Prepackaged Software  
1      nan        Health Care                       Medical/Dental Instruments  
2  2013.00  Consumer Services            Services-Misc. Amusement & Recreation  
3  2013.00         Technology  Computer Software: Programming, Data Processing  
4      nan         Technology                                     EDP Services

In [27]:
# run and inspect!
all_listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 9 columns):
Exchange     6852 non-null object
Symbol       6852 non-null object
Name         6852 non-null object
LastSale     6745 non-null float64
MarketCap    6852 non-null float64
ADR TSO      140 non-null float64
IPOyear      3105 non-null float64
Sector       5309 non-null object
Industry     5309 non-null object
dtypes: float64(4), object(5)
memory usage: 481.9+ KB


There are in total ... 6852 listings!

# Well Done!

---------------------------------

# Hints (Spoiler!)

76. drop() method, columns parameter

77. listings.Exchange != "AMEX"

78. pass "MCAP_MUSD" as new column label, Market_Cap column

79. use the div() method -> listings[---] = listings.---.div(listings.---)

80. sort_values() method

81. run the cells and inspects!

82. pd.concat() method, keys parameter